src/bp/chain.rs

```diff
@@ -0,0 +1,1413 @@
+            P2pNetworkId::Other(n) => *n,
+        }
+    }
+}
+
+impl StrictEncode for P2pNetworkId {
+    type Error = strict_encoding::Error;
+
+    #[inline]
+    fn strict_encode<E: io::Write>(&self, e: E) -> Result<usize, Self::Error> {
+        Ok(self.as_magic().strict_encode(e)?)
+    }
+}
+
+impl StrictDecode for P2pNetworkId {
+    type Error = strict_encoding::Error;
+
+    #[inline]
+    fn strict_decode<D: io::Read>(d: D) -> Result<Self, Self::Error> {
+        Ok(Self::from_magic(u32::strict_decode(d)?))
+    }
+}
+
+impl Debug for P2pNetworkId {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{}({:#x?})", self, self.as_magic())
+    }
+}
+
+impl Display for P2pNetworkId {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            P2pNetworkId::Mainnet => f.write_str("mainnet"),
+            P2pNetworkId::Testnet => f.write_str("testnet"),
+            P2pNetworkId::Regtest => f.write_str("regtest"),
+            P2pNetworkId::Signet => f.write_str("signet"),
+            P2pNetworkId::Other(_) => f.write_str("unknown"),
+        }
+    }
+}
+
+impl From<P2pMagicNumber> for P2pNetworkId {
+    fn from(magic: P2pMagicNumber) -> Self {
+        P2pNetworkId::from_magic(magic)
+    }
+}
+
+impl From<P2pNetworkId> for P2pMagicNumber {
+    fn from(network: P2pNetworkId) -> Self {
+        network.as_magic()
+    }
+}
+
+impl From<bitcoin::Network> for P2pNetworkId {
+    fn from(bn: bitcoin::Network) -> Self {
+        match bn {
+            bitcoin::Network::Bitcoin => P2pNetworkId::Mainnet,
+            bitcoin::Network::Testnet => P2pNetworkId::Testnet,
+            bitcoin::Network::Regtest => P2pNetworkId::Regtest,
+            bitcoin::Network::Signet => P2pNetworkId::Signet,
+        }
+    }
+}
+
+impl TryFrom<P2pNetworkId> for bitcoin::Network {
+    type Error = NoneError;
+    fn try_from(bn: P2pNetworkId) -> Result<Self, Self::Error> {
+        Ok(match bn {
+            P2pNetworkId::Mainnet => bitcoin::Network::Bitcoin,
+            P2pNetworkId::Testnet => bitcoin::Network::Testnet,
+            P2pNetworkId::Regtest => bitcoin::Network::Regtest,
+            P2pNetworkId::Signet => bitcoin::Network::Signet,
+            P2pNetworkId::Other(magic) if magic == P2P_MAGIC_MAINNET => bitcoin::Network::Bitcoin,
+            P2pNetworkId::Other(magic) if magic == P2P_MAGIC_TESTNET => bitcoin::Network::Testnet,
+            P2pNetworkId::Other(magic) if magic == P2P_MAGIC_REGTEST => bitcoin::Network::Regtest,
+            P2pNetworkId::Other(magic) if magic == P2P_MAGIC_SIGNET => bitcoin::Network::Signet,
+            _ => Err(NoneError)?,
+        })
+    }
+}
+
+hash_newtype!(
+    AssetId,
+    sha256d::Hash,
+    32,
+    doc = "Universal asset identifier for on-chain and off-chain assets; for \
+           on-chain assets matches genesis hash of the chain, but displayed in \
+           normal, non-reverse order",
```

Comment:
I am a little confused by the wording here. Will the `AssetId` **match** with genesis chain hash? Or it will **commit** to the genesis chain hash?